group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -0,0 +1,535 @@
+        Optional<String> serverAssignorNameOpt
+    ) {
+        Map<String, Integer> counts = new HashMap<>();
+
+        serverAssignorNameOpt.ifPresent(serverAssignorName ->
+            counts.put(serverAssignorName, 1)
+        );
+
+        members.forEach((memberId, member) -> {
+            if (!memberId.equals(updatedMemberId) && member.serverAssignorName().isPresent()) {
+                counts.compute(member.serverAssignorName().get(), (k, v) -> v == null ? 1 : v + 1);
+            }
+        });
+
+        return counts.entrySet().stream()
+            .max(Map.Entry.comparingByValue())
+            .map(Map.Entry::getKey);
+    }
+
+    /**
+     * Returns the subscription metadata for all the topics whose
+     * members are subscribed to.
+     *
+     * @return An immutable Map containing the subscription metadata.
+     */
+    public Map<String, TopicMetadata> subscriptionMetadata() {
+        return Collections.unmodifiableMap(subscribedTopicMetadata);
+    }
+
+    /**
+     * Updates the subscription metadata. This replace the previous one.
+     *
+     * @param subscriptionMetadata The new subscription metadata.
+     */
+    public void setSubscriptionMetadata(
+        Map<String, TopicMetadata> subscriptionMetadata
+    ) {
+        this.subscribedTopicMetadata.clear();
+        this.subscribedTopicMetadata.putAll(subscriptionMetadata);
+    }
+
+    /**
+     * Computes new subscription metadata but with specific information for
+     * a member.
+     *
+     * @param memberId                      The member id.
+     * @param updatedMemberSubscriptions    The member's updated topic subscriptions.
+     * @param topicsImage                   The topic metadata.
+     *
+     * @return The new subscription metadata as an immutable Map.
+     */
+    public Map<String, TopicMetadata> computeSubscriptionMetadata(
+        String memberId,
+        List<String> updatedMemberSubscriptions,
+        TopicsImage topicsImage
+    ) {
+        Map<String, TopicMetadata> newSubscriptionMetadata = new HashMap<>(subscriptionMetadata().size());
+
+        Consumer<List<String>> updateSubscription = subscribedTopicNames -> {
+            subscribedTopicNames.forEach(topicName ->
+                newSubscriptionMetadata.computeIfAbsent(topicName, __ -> {
+                    TopicImage topicImage = topicsImage.getTopic(topicName);
+                    if (topicImage == null) {
+                        return null;
+                    } else {
+                        return new TopicMetadata(
+                            topicImage.id(),
+                            topicImage.name(),
+                            topicImage.partitions().size()
+                        );
+                    }
+                })
+            );
+        };
+
+        if (updatedMemberSubscriptions != null) {
+            updateSubscription.accept(updatedMemberSubscriptions);
+        }
+
+        members.forEach((mid, member) -> {
```

Comment:
I think that keeping a map of TopicMetadata to member ids does not work because the content of TopicMetadata could change as well. However, we could keep a map of topic name to number of subscribers. That would reduce the computation. I did the change. Let me know what you think.